File: cpplearn/crtp/obj_cnt_crtp.py
class ObjectCounter:
    counter = 0
    def __init__(self):
        type(self).counter +=1
    def __del__(self):
        type(self).counter -= 1

    @classmethod
    def count_live(cls):
        return cls.counter
    
class MyVector(ObjectCounter):
    pass

class MyCharStr(ObjectCounter):
    pass

v1, v2 = MyVector(), MyVector()
c1 = MyCharStr()
print(MyVector.count_live()) # 2
print(MyCharStr.count_live()) # 1

File: cpplearn/crtp/obj_cnt_crtp.cpp
#include <vector>
#include <format>
#include <iostream>

template <class Derived>
class ObjectCounter {
private:
	static inline size_t count{ 0 };

protected:
	ObjectCounter() { ++count; }
	ObjectCounter(const ObjectCounter&) { ++count; }
	ObjectCounter(ObjectCounter&&) { ++count; }
	~ObjectCounter() { --count; }

public:
	static std::size_t CountLive() { return count; }
};

template <typename T>
class MyVector : public ObjectCounter<MyVector<T>> {};

class MyCharString : public ObjectCounter<MyCharString> {};

int main() {
	MyVector<int> v1, v2;
	MyCharString s1;


	std::cout <<std::format("number of MyVector: {}\n", MyVector<int>::CountLive());
	std::cout<<std::format("number of myCharString: {}\n", MyCharString::CountLive());
	return 0;
}

